Adding hd44780 driver

dhrone opened the pull request:
Requesting a draft PR as I need some guidance on documentation.  Currently the luma.lcd docs only supply information on SPI style interfaces which are included on the install.rst.  There is no hardware section to describe different interfaces types like luma.oled has.  The hd44780 driver supports the new parallel and pcf8574 interfaces so I can either place the information for interfacing with the hd44780 on a dedicated page for it, or modify the luma.lcd docs to be more consistent with the luma.oled ones by adding a hardware.rst and modifying the install.rst.  That's the path I would suggest but it would mean largely copying hardware.rst from luma.oled to luma.lcd.  Either way, I plan to add a page to the document dedicated to the hd44780 to explain how the character and font functions work for it.  Let me know what direction you'd like taken.

thijstriemstra commented on luma/lcd/device.py:
```diff
@@ -757,3 +760,234 @@ def contrast(self, value):
+    .. versionadded:: 2.5.0
+    """
+    def __init__(self, serial_interface=None, width=16, height=2, undefined='_',
+                 selected_font=0, exec_time=1e-6 * 50, framebuffer="diff_to_previous", **kwargs):
```
value for exec_time is a constant in luma.core now, can you use that here?

thijstriemstra replied:
Oh PULSE_TIME isn't part of `__all__` though. Maybe it should be moved to `__all__` or a different place so it can be re-used in sub-projects.

dhrone replied:
Actually pulse_time and exec_time are different values.  pulse_time sets the duration of the 'enable' pulse (low-high-low) on the enable pin of a parallel-bus display.  Exec_time is to give the display time to process a command before it receives another message.  

In recent testing, I've been examining the lower-limit for using sleep to create these delays.  At least on RPis using Raspbian, it appears that the smallest delay you can request is 1.2e-5.  So the 1e-6 isn't really giving me the correct timing anyway.  Eventually, I'll look into seeing if there is another method that I can use to get closer to the spec sheets for the displays but for now, slower still works.

As far as correct placement in the code, each parallel device has its own requirements but 1e-5 is likely to be above the required value for pretty much any display.  So, leaving it in luma.core.interface.serial makes sense.

thijstriemstra replied:
sounds good.

dhrone replied:
I went down a rabbit hole trying to get better GPIO performance by writing a c++ class using libgpiod.  Bottom line is even at the speeds I could get out of that library, the required pulse_time for the ws0010 is shorter than the time libgpiod takes to move the pin from low to high to low.  So effectively, the default pulse_time for the ws0010 (and winstar_weh) can be set to 0.  This significantly improves the update time for that display.  However, if you are trying to horizontally scroll anything, the upper limit appears to be around 30 FPS.  Above that, animations show tearing.  

I haven't done as much testing of the hd44780 displays but they appear to be much slower devices.  Setting pulse_time to zero does cause stability problems so we are stuck using at least some delay.  For now, I'll just leave the default to 1e-6 but may want to change from using sleep to a timing loop in future versions.  Also, 5 FPS was the sweet spot for refreshing it.

thijstriemstra commented:
> modify the luma.lcd docs to be more consistent with the luma.oled ones by adding a hardware.rst and modifying the install.rst. 

+1

dhrone commented:
Ok.  I'll get to modifying the docs accordingly.  QQ: Did a release go out for luma.core?  I'm still seeing pypy showing 1.16.0 for it.

rm-hull commented:
Not yet, I’ll try and get release 1.16.1 out this evening

dhrone commented:
Thanks @rm-hull and sorry for the trouble.  I got tripped up testing across the projects.

rm-hull commented:
@dhrone luma.core 1.16.1 was released

thijstriemstra commented on tests/test_hd44780.py:
```diff
@@ -0,0 +1,187 @@
+from PIL import Image, ImageDraw
+from unittest.mock import Mock, call
+
+CLEAR = 0x01
```
can these constants be moved to a temporary test class/enum and ref them like that.

dhrone replied:
They were unnecessary.  I referenced the version of the constants from luma/lcd/const.py

thijstriemstra commented on tests/test_hd44780.py:
```diff
@@ -0,0 +1,187 @@
+
+def test_get_font():
+    """
+    Verify get font capability by requesting two fonts and printing a single
```
No need to use word "Verify". I'd omit it, tests are always verifying.

dhrone replied:
I've eliminated the use of the word verify.

thijstriemstra commented on tests/test_hd44780.py:
```diff
@@ -0,0 +1,187 @@
+def test_custom_full():
+    """
+    Test of behavior when auto-create feature runs out of custom character
+    space
```
~~Test of behavior when~~

dhrone replied:
Changed as you have suggested

thijstriemstra commented on luma/lcd/device.py:
```diff
@@ -757,3 +760,234 @@ def contrast(self, value):
+            Most displays have limited memory to support custom characters
+            and typically can only support 8 at any one time.  If this is
+            exceeded, the remaining unmatched characters will be replaced by
+            the undefined character.
```
undefined should probaby be in double backticks? ```undefined```

dhrone replied:
Changed as suggested

thijstriemstra commented on luma/lcd/device.py:
```diff
@@ -757,3 +760,234 @@ def contrast(self, value):
+        ``diff_to_previous`` or ``full_frame`` are only supported.
+    :type framebuffer: str
+
+    To place text on the display, simply assign the text to the 'text'
```
instead of 'text' use double backticks for ```text```

dhrone replied:
Changed as suggested

thijstriemstra commented on luma/lcd/device.py:
```diff
@@ -757,3 +760,234 @@ def contrast(self, value):
+        font tables
+    :type undefined: str
+    :param selected_font: the font table appropriate for the model of display you
+        are using.  The hd44780 normally comes in a version with font A00 (ENGLISH_JAPANESE) or A02 (ENGLISH_EUROPEAN).  You can provide either
```
break this line on around 80-100 chars (as well as any others).

dhrone replied:
Changed as suggested

thijstriemstra commented on doc/index.rst:
```diff
@@ -18,12 +18,13 @@ Luma.LCD: Display drivers for PCD8544, ST7735, ST7567, HT1621, UC1701X, ILI9341
    :maxdepth: 2
 
    intro
-   install
```
having install as high up in list as possible has my preference.

dhrone replied:
Moved install up to the second section (right after intro).  The only issue is need to make sure that the hardware section requisites have been completed but no harm in having hardware a bit later.

thijstriemstra commented on doc/install.rst:
```diff
@@ -1,7 +1,10 @@
-.. note:: The library has been tested against Python 2.7, 3.4, 3.5 and 3.6.
+.. note:: The library has been tested against Python 3.5, 3.6, 3.7 and 3.8.
 
    For **Python3** installation, substitute the following in the
```
since python3 is only supported, no need for this section anymore (and rest of doc has to be updated for py3).

dhrone replied:
Removed reference to 2.7 and 3.4.  Added, 3.7 and 3.8.  Removed section that reference the versions of pip, python, etc.

thijstriemstra commented on doc/install.rst:
```diff
@@ -16,173 +19,7 @@ Installation
    (both Raspbian Jessie and Stretch) models A, B2, 3B, Zero, Zero W and
```
Maybe leave out the distro names, since we have to update it with every release.

dhrone replied:
Since it references specific work that you have done, perhaps it should remain as is.  I would suggest eventually including a dedicated page on test results.

thijstriemstra commented on doc/install.rst:
```diff
@@ -196,9 +33,8 @@ with::
-   cause components to not be installed properly. Please ensure that **pip 9.0.1** 
-   is installed prior to continuing::
-   
+.. warning:: The default pip bundled with apt on Raspbian Jessie is really old,
```
This warning can be removed, Jessie is ancient.

dhrone replied:
Removed

thijstriemstra commented on doc/intro.rst:
```diff
@@ -11,6 +11,12 @@ canvas, and other functionality to support:
 * color/greyscale (where supported),
 * dithering to monochrome
 
+**New**  The HD44780 (and similar) devices are some of the most popular small LCD displays
```
I would remove the 'new' because it'll become outdated.

dhrone replied:
Just a bit of marketing, but I agree.  It has been removed.

thijstriemstra commented on doc/python-usage.rst:
```diff
@@ -1,37 +1,34 @@
 Python usage
 ------------
+LCD displays can be driven with python using the various implementations in the :py:mod:`luma.lcd.device` package.  There are several device classes available and usage is very simple if you have ever use `Pillow <https://pillow.readthedocs.io/en/latest/>`_ of PIL.
```
try to wrap lines between 80 and 100.

dhrone replied:
Fixed

thijstriemstra commented on luma/lcd/device.py:
```diff
@@ -89,6 +89,50 @@ def __call__(self, is_enabled):
+        """
+        Toggle the LCD Backlight
+
+        :param is_enabled: Turn on or off the backlight.
```
'Turn backlight on or off'

dhrone replied:
Changed as suggested

thijstriemstra replied:
You changed it elsewhere, not here yet.

thijstriemstra commented on luma/lcd/device.py:
```diff
@@ -139,16 +183,21 @@ class backlit_device(device):
     .. versionadded:: 2.0.0
     """
-    def __init__(self, const=None, serial_interface=None, gpio=None, gpio_LIGHT=18, active_low=True, pwm_frequency=None, **kwargs):
+    def __init__(self, const=None, serial_interface=None, gpio=None, gpio_LIGHT=18, active_low=True, pwm_frequency=None, backpack_pin=None, **kwargs):
```
try wrap at 80

dhrone replied:
Changed as suggested

thijstriemstra commented on doc/HD44780.rst:
```diff
@@ -0,0 +1,197 @@
+for details on which characters are included in each.
+
+In addition to the built-in characters, HD44780s can have up to eight custom
+written into their RAM.  This feature is used by Luma.LCD to provide a limited
```
i think we name it using all lowercase, so: luma.lcd

dhrone replied:
There is some small inconsistencies in the documentation on that point but I've adjusted this plus another instance in intro.rst.

thijstriemstra commented on doc/HD44780.rst:
```diff
@@ -0,0 +1,197 @@
+  from luma.core.interface.serial import parallel
+  from luma.lcd.device import hd44780
+
+  p = parallel(RS=7, E=8, PINS=[25,24,23,27])
```
no need to use single char vars, use something descriptive instead.

dhrone replied:
They are for the (R)egister(S)elect pin and the (E)nable pin.  Anyone wiring these displays should know what these mean and the naming convention is consistent with the ht1621 class.  Take a look and if you still want them changed, I'll rename them.

rm-hull replied:
Was @thijstriemstra talking about `p` (for the parallel interface) and `d` (for the device) ?

As an aside, I'm also wondering if `parallel` is  in the right package (`from luma.core.interface.serial import parallel`) and whether it is a good name for the interface ... perhaps it aught be something like `from luma.core.interface.parallel import bit_bang` ?

dhrone replied:
Ahh, I misunderstood.  I can adjust the variable names in the example.  

To your other point, I can certainly move the parallel class into its own module within luma.lcd but I'm wondering if there is a different way that the interfaces should be organized.  The common interfaces I am aware of are...

* Parallel 6800
* Parallel 8080
* SPI 3 wire
* SPI 4 wire
* I2C

In addition, I2C backpacks such as the PCF8574, MCP23008 and the MCP23017 are often used with the I2C interface to bitbang devices that only support a parallel interface. 

Perhaps collapsing the serial interfaces into the interface module instead of considering them their own interface type.

luma.core.interface
* SPI
* I2C
* PCF8574
* MCP23008
* MCP23017
* ftdi_spi
* ftdi_i2c
* bitbang_6800
* bitbang_8080
* bitbang_spi
* bitbang_serial (for devices like the ht1621)

thijstriemstra commented on doc/HD44780.rst:
```diff
@@ -0,0 +1,197 @@
+
+There are a few of things that deserve highlighting in this code.
+
+* We have used the hd44780 classes get_font method to retrieve the internal
```
get_font needs backticks

dhrone replied:
I've placed double backticks around the get_font reference.

thijstriemstra commented on doc/HD44780.rst:
```diff
@@ -0,0 +1,197 @@
+  font tables.  You should request the table that matches what is installed in
+  your display.
+
+* The progress bar is drawn using normal ImageDraw primitives in this case a
```
ImageDraw should link to pillow api docs.

dhrone replied:
Changed as suggested.

thijstriemstra commented on doc/HD44780.rst:
```diff
@@ -0,0 +1,197 @@
+A02          English European
+===========  ==================
+
+You can retrieve either of these fonts using the `get_font` method.
```
use double backticks (or api link) instead of single

dhrone replied:
Changed to double backticks

thijstriemstra commented on doc/HD44780.rst:
```diff
@@ -0,0 +1,197 @@
+You can also combine fonts together in order to display characters not included
+within your device's character table.
+
+As an example, the 'Black Right-Pointing Triable' symbol (Unicode U+25b6) is not
```
i would put the unicode char in backticks

dhrone replied:
Added double backticks and removed parens.

thijstriemstra commented on doc/HD44780.rst:
```diff
@@ -0,0 +1,197 @@
+  device.text = '\u25b6 Play'
+
+This feature leverages the custom character capability so it has the same 8
+character limitation.  If you exceed 8 characters within a screen, the `undefined`
```
undefined needs double backticks

dhrone replied:
Added double backticks

thijstriemstra commented on doc/HD44780.rst:
```diff
@@ -0,0 +1,197 @@
+within the devices font table.
+
+See the documentation for :py:mod:`luma.core.bitmap_font` for more information on
+how to use the bitmap_font module.
```
bitmap_font needs double backticks

dhrone replied:
Added double backticks

thijstriemstra commented on doc/hardware.rst:
```diff
@@ -0,0 +1,426 @@
+
+Then reboot.
+
+Next, add your user to the *i2c* group and install ``i2c-tools``::
```
i would use double backticks for i2c

dhrone replied:
This was copied from the luma.oled content.  If we change it here, probably should change it in both places.

thijstriemstra replied:
make sense (to start here at least)

thijstriemstra commented on doc/hardware.rst:
```diff
@@ -0,0 +1,426 @@
+    :py:attr:`port=1` in your serial interface create call.
+
+The GPIO pins used for this SPI connection are the same for all versions of the
+Raspberry Pi, up to and including the Raspberry Pi 3 B.
```
It's probably safe to change this to 4.

dhrone replied:
Same comment as above but I did make the change for this one.

thijstriemstra commented on doc/hardware.rst:
```diff
@@ -0,0 +1,426 @@
+  $ lsmod | grep spi
+  spi_bcm2835             6678  0
+
+Then add your user to the *spi* and *gpio* groups::
```
doubleback ticks for spi and gpio

dhrone replied:
Consistency issue as above.

thijstriemstra replied:
i dont mind consistency doc issues between core and sub-projects. At some point it'll be fixed in both..

thijstriemstra commented on doc/hardware.rst:
```diff
@@ -0,0 +1,426 @@
+""""""""""""""""""""""""
+
+.. note:
+  The please review the :py:class:`luma.core.interface.serial.pcf8574`
```
remove 'The'

dhrone replied:
Fixed as suggested

thijstriemstra commented on doc/install.rst:
```diff
@@ -1,188 +1,16 @@
 
    It was *originally* tested with Raspbian on a rev.2 model B, with a vanilla
    kernel version 4.1.16+, and has subsequently been tested on Raspberry Pi
    (both Raspbian Jessie and Stretch) models A, B2, 3B, Zero, Zero W and
```
add 4 here as well.

dhrone replied:
Added reference to 4B

thijstriemstra commented on doc/python-usage.rst:
```diff
@@ -1,37 +1,32 @@
 ------------
+LCD displays can be driven with python using the various implementations in the
+:py:mod:`luma.lcd.device` package.  There are several device classes available
+and usage is very simple if you have ever use
```
use -> used

dhrone replied:
Changed as suggested

thijstriemstra commented on doc/python-usage.rst:
```diff
@@ -1,37 +1,32 @@
+LCD displays can be driven with python using the various implementations in the
+:py:mod:`luma.lcd.device` package.  There are several device classes available
+and usage is very simple if you have ever use
+`Pillow <https://pillow.readthedocs.io/en/latest/>`_ of PIL.
```
of -> or

dhrone replied:
Changed as suggested

thijstriemstra commented on doc/python-usage.rst:
```diff
@@ -1,37 +1,32 @@
-The display device should now be configured for use. Note, all the example code
-snippets in this section are interchangeable between PCD8544 and ST7735
-devices.
+The display device should now be configured for use.
```
configured for use. -> properly configured.

dhrone replied:
Changed as suggested but consistency issue with luma.oled docs.

thijstriemstra commented on doc/python-usage.rst:
```diff
@@ -49,10 +44,15 @@ As soon as the with scope is ended, the resultant image is automatically
 
+.. note::
+  The use of the display method for the HD44780 is more limited than the other
+  LCDs.  For it, the `text` property is the preferred interface for displaying
```
Remove 'For it,'

dhrone replied:
Changed as suggested

thijstriemstra commented on doc/python-usage.rst:
```diff
@@ -49,10 +44,15 @@ As soon as the with scope is ended, the resultant image is automatically
-Any of the standard :py:mod:`PIL.ImageColor` color formats may be used, but
-since the PCD8544 LCD is monochrome, only the HTML color names
+Any of the standard :py:mod:`PIL.ImageColor` color formats may be used if your
+device support them.  For monochrome LCDs, only the HTML color names
```
support -> supports

dhrone replied:
Changed as suggested

thijstriemstra commented on doc/python-usage.rst:
```diff
@@ -64,22 +64,23 @@ effect (see the *3d_box.py* example, below).
 
-Note that there is no such limitation for the ST7735 or ILI9341 devices which supports 262K
-colour RGB images, whereby 24-bit RGB images are downscaled to 18-bit RGB.
+The ST7735 and ILI9341 devices can display 262K colour RGB images.  If supplied
```
If supplied -> When supplying

dhrone replied:
Changed as suggested

thijstriemstra commented on doc/python-usage.rst:
```diff
@@ -64,22 +64,23 @@ effect (see the *3d_box.py* example, below).
-an application that requires the display to be mounted in a portrait aspect,
-then add a :py:attr:`rotate=N` parameter when creating the device:
+By default the PCD8544, ST7735, UC1701X and ILI9341 displays will all be oriented
+inlandscape mode (84x48, 160x128, 128x64 and 320x240 pixels respectively). Should
```
in landscape

dhrone replied:
Added needed space

thijstriemstra commented on doc/python-usage.rst:
```diff
@@ -139,12 +142,25 @@ buffer allows, but only because dots are folded into their host character.
-device. The backlight can be programmatically switched on and off by calling
+display.  If the display's backlight is connected to one of the single-board
+computer's gpio pins, you can activate the backlight by specifying
+``gpio_LIGHT=n`` where n = the pin number when initializing the
```
double backticks for n

dhrone replied:
Changed as suggested

thijstriemstra commented on doc/python-usage.rst:
```diff
@@ -139,12 +142,25 @@ buffer allows, but only because dots are folded into their host character.
-repository. Details of how to run the examples is shown in the example repo's README.
+After installing the library, head over to the
+`luma.examples <https://github.com/rm-hull/luma.examples>`_
+repository. Details of how to run the examples is shown in the example repo's
```
example repo's README sounds complicated. Maybe just 'README' and link to it directly?

dhrone replied:
Changed as suggested.

thijstriemstra commented on luma/lcd/device.py:
```diff
@@ -757,3 +809,247 @@ def contrast(self, value):
+            self.command(0x30, exec_time=1e-6 * 200)
+            self.command(0x30)
+        else:
+            ''' In case of stubborn display place in 8 bit mode three times and
```
use # instead of '''

dhrone replied:
Changed to # comments as suggested

thijstriemstra commented on luma/lcd/device.py:
```diff
@@ -757,3 +809,247 @@ def contrast(self, value):
+        :type image: :py:class:`PIL.Image.Image`
+
+        If needed, it will create custom characters if a glyph is not found
+        within the font table
```
missing period.

dhrone replied:
Added period

thijstriemstra commented on tests/test_hd44780.py:
```diff
@@ -0,0 +1,203 @@
+
+def test_i2c_does_not_support_backlight():
+    """
+    Test exception is thrown if supplied serial_interface does not support a backlight
```
Test -> An

dhrone replied:
Changed as suggested

thijstriemstra commented on tests/test_hd44780.py:
```diff
@@ -0,0 +1,203 @@
+
+def test_unsupported_display_mode():
+    """
+    Test exception is thrown if an unsupported display mode is requested
```
Test -> An

dhrone replied:
Changed as suggested

rm-hull commented on doc/HD44780.rst:
```diff
@@ -0,0 +1,197 @@
+  from luma.core.render import canvas
+  from PIL import Image, ImageDraw
+
+  serial = parallel(RS=7, E=8, PINS=[25,24,23,27])
```
Potentially confusing name for a parallel device

dhrone replied:
Changed the examples in HD44780.rst to use less confusing variable names.

rm-hull commented on doc/hardware.rst:
```diff
@@ -0,0 +1,426 @@
+Hardware
+--------
+
+Identifying your serial interface
```
Identifying your ~~serial~~ interface

dhrone replied:
Removed serial and fixed a grammar issue in the sentence about the number of pins in a parallel interfaced device.

rm-hull reviewed (approved):
Looks fine, apart from comments noted inline. I would like us to have a discussion around the package names for the parallel interface a bit though before this is released

dhrone commented:
I suggested an organization around ``luma.core.interface`` above.  However, I acknowledge the change may not be worth the impact. The simplest change seems to be to create a parallel module within ``luma.core.interface``.

``luma.core.interface.serial``
* SPI
* I2C
* ftdi_spi
* ftdi_i2c
* PCF8574
* MCP23008 (TBD)
* MCP23017 (TBD)
* bitbang_spi
* bitbang_serial (TBD - for ht1621 style interfaces)

``luma.core.interface.parallel``
* bitbang_6800
* bitbang_8080 (TBD)

dhrone commented:
I have an unrelated question.  During testing for the ws0010, I wound up writing a C++ version of ``luma.core.parallel``.  It leverages the new chardev interface introduced in Linux kernel 4.8 and the libgpiod library (v1.4.4).  Performance tests seem to show around a 20-25% advantage over the python version at least on an RPi 3b.  The libgpiod is under active development and 1.5.2 is the current version but I am using the older one as 1.5 required Linux Kernel 5.5+ which is not currently released for most RPi focused distros.  I chose to use the chardev interface as it should be stable across any SBCs that are running the 4.8 or better Linux kernel (e.g. not RPi specific).  

So my question is, do you want to introduce libgpiod as the way that the GPIO is controlled?  I do have the C++ library interfaced to python using ctypes but I unfortunately have no experience adding the compiling of the c code into a python (setup.py) build.  I would need your assistance if you decided to use it.

rm-hull commented:
In regards the C++ version, if it can be installed painlessly by end-users (and by painlessly, I mean doesn't introduce lots of _"It doesnt work for me"_-type issues being created) then I don't see why not... however the only sticky thing might be how to test it via the travis CI runner. We have currently used the python unittest.mocks to patch the RPI.GPIO library (if my memory serves correctly) and this seemed to work reasonably well. If using a C++ module and relying on libgpiod means it is baking in a hard dependency that makes it difficult or impossible to test, this might be a deal-breaker.

Although having said all that... when you say it is 25% faster, have you got any hard numbers? If it can already support a frame-rate of say 200FPS for example on 16x2 character device, and the C++ lib boosts it to 250 FPS then I would say its not worth it (because the frame rate is already very high), but if it is say 4 FPS and it goes up to 5 FPS then it may be worth it

dhrone commented:
Hold up on publishing this PR.  @satiromarra discovered a GPIO cleanup bug in the backlight code.  I'm chasing down what the correct clean-up approach should be.

dhrone commented:
Getting back to the libgpiod question.  It really depends on the display.  For the two I am testing with (winstar weg010016a, hd44780 16x2), it really wouldn't be worth the effort.  While I can physically get the data to the displays faster with the library, the amount of time it takes the displays to paint their screens is the limiting factor.  For the hd44780s anything above about 5 fps looks horrible.  For the weg010016a its about 30 fps.  Above that you get tearing and other undesirable display artifacts.  If you had a larger display, say 256x128, it might be worth it if the device doesn't support SPI.

thijstriemstra commented:
> In regards the C++ version, if it can be installed painlessly by end-users (and by painlessly, I mean doesn't introduce lots of "It doesnt work for me"-type issues being created) then I don't see why not... 

There's some pi-specific registery enabled by default in raspbian that will pre-compile native bindings for rpi. I hope it'll automatically do it for luma.* but not sure how it works (maybe preference based on popularity?)

thijstriemstra commented:
> I suggested an organization around luma.core.interface above.

That's unfortunate, but it happens, and I would suggest a luma.core 2.0 for this.

dhrone commented:
For the organization of ``luma.core.interface``, if we use the approach I laid out in my last message, the only immediate change would be the movement and renaming of the current luma.core.interface.serial.parallel class to ``luma.core.interface.parallel.bitbang_6800``.   To prevent that from being breaking for the recent release of ``luma.oled`` we could cross import ``bitbang_6800`` back to ``luma.core.interface.serial.parallel``.   I'd recommend we leave the discussion around the new drivers I suggested for a later release.  Thoughts?

dhrone commented:
@rm-hull Is there anything else you'd like to see before a merge and release?

rm-hull commented:
Yep I'm happy for it to be merged if @thijstriemstra is good for approval

thijstriemstra commented:
> To prevent that from being breaking for the recent release of luma.oled

It wasn't released yet.

> s there anything else you'd like to see before a merge and release?

Maybe make the change to core first and then change/release luma.lcd?

thijstriemstra commented on tests/test_ht1621.py:
```diff
@@ -107,7 +107,7 @@ def reset_mock(self):
         self._state = 'none'   # Current state
 
-    def cleanup(self):
+    def cleanup(self, *args):
```
this change isn't needed is it?

dhrone replied:
I think it is needed.  The problem is the ``MockHT1621`` class is intended to be a test object for the ``RPi.GPIO`` class and ``RPi.GPIO.cleanup`` optionally accepts an integer or a list of integers.  I am now using that interface to selectively clean up the GPIO pin used by the GPIO and PWM backlight classes within luma.lcd.  See luma.lcd.device.py:93 and 180.

dhrone commented:
Quick update.   Items in flight but unsubmitted:
* luma.lcd and luma.oled documentations edited to make them consistent
* new parallel module added to luma.core.interface
* moved parallel class from luma.core.interface.serial, renamed to bitbang_6800 and placed in luma.core.interface.parallel

To do:
* Update luma.core documentation to reflect new module/class.
* Update tests to ensure bitbang_6800 functions correctly
* Ensure changes still pass all functional, quality, and documentation checks
* Issue PRs for all three projects

I'll try to get the to-do's done this weekend.  

Is there anything I'm missing?

dhrone commented:
Merging hd44780 branch in main

thijstriemstra commented:
@rm-hull can you also add HD44780 to the github description (and any others if missing)?
